Improved routing key assignment with graph colouring

mundya opened the pull request:
This improves (and builds upon) #133 by allowing the same keys to be used by signals which target the same filter models in the same operators. The result is that yet fewer unique signal identifiers are used and that the filter routing table gets smaller, reducing the compute cost of receiving and processing a multicast packet.
- [x] More testing of this with larger and complex models.

mundya committed:
Request signal ID constraints from operators

Query operators to determine which of the signals that feed them cannot
share a key. Should the operator not contain the method
`get_signal_id_constraints` then it is assumed that all signals arriving
at the operator should be uniquely identified.

mundya commented on nengo_spinnaker/builder/builder.py:
```diff
@@ -399,8 +416,15 @@ def make_netlist(self, *args, **kwargs):
+
+        # Get the constraints on the signal identifiers
+        signal_id_constraints = dict()
+        for u, vs in id_constraints:
```
Should be `for u, vs in iteritems(id_constraints):`

neworderofjamie commented on nengo_spinnaker/netlist/key_allocation.py:
```diff
@@ -113,9 +87,9 @@ def build_mn_net_graph(nets_routes, prior_constraints=None):
-            # Add an edge iff. it would connect two *different* vertices
-            if x != y:
+        for xs, ys in itertools.combinations(itervalues(route_nets), 2):
+            for x, y in ((i, j) for j in ys for i in xs if i != j):
```
oh god - I hope that this can indeed be replaced by combinations

neworderofjamie commented on nengo_spinnaker/regions/filters.py:
```diff
@@ -318,6 +319,36 @@ def sizeof(self, *args):
+        # The signals in different target-sets cannot share routing identifiers
+        constraints = collections.defaultdict(set)
+        for xs, ys in combinations(itervalues(targets_to_ids), 2):
+            for x, y in ((i, j) for j in ys for i in xs if i != j):
```
again, I really hope this can be replaced with something combinationsish

mundya committed:
Fixes and neatening

Thanks @neworderofjamie

mundya commented:
I'll probably merge this tomorrow, just want a bit more time to test it with larger models.

mundya committed:
Key allocation for ensembles

Allows the graph colouring key allocation to be used for signals
arriving at ensembles. This boils down to the constraints on signals
being the union of the constraints required by the filter routing
regions AND the constraints required to ensure that signals which target
different sets of routing regions do not share a routing key.

In addition a small fix to the reporting of the routing key identifiers
and a fix to ensure that signals which have already been assigned
keyspaces are not included in the set of constraints reported by the
filter routing regions.

mundya commented:
@neworderofjamie can I do a review of the latest commit with you please?

neworderofjamie commented:
Yep - looks good
